File: sdks/go/didcomm.go
package atap

import (
	"context"
	"fmt"
)

// DIDCommAPI provides DIDComm messaging operations.
type DIDCommAPI struct {
	client *Client
}

// Send sends a DIDComm message (JWE envelope). Public endpoint.
func (d *DIDCommAPI) Send(ctx context.Context, jweBytes []byte) (map[string]interface{}, error) {
	data, err := d.client.http.Request(ctx, "POST", "/v1/didcomm", &RequestOptions{
		RawBody:     jweBytes,
		ContentType: "application/didcomm-encrypted+json",
	})
	if err != nil {
		return nil, fmt.Errorf("send didcomm message: %w", err)
	}
	return data, nil
}

// Inbox retrieves pending DIDComm messages. Requires atap:inbox scope.
func (d *DIDCommAPI) Inbox(ctx context.Context, limit int) (*DIDCommInbox, error) {
	if limit <= 0 {
		limit = 50
	}
	if limit > 100 {
		limit = 100
	}

	data, err := d.client.authedRequest(ctx, "GET", "/v1/didcomm/inbox", &RequestOptions{
		Params: map[string]string{"limit": fmt.Sprintf("%d", limit)},
	})
	if err != nil {
		return nil, err
	}

	var messages []DIDCommMessage
	if items, ok := data["messages"].([]interface{}); ok {
		for _, item := range items {
			if m, ok := item.(map[string]interface{}); ok {
				messages = append(messages, DIDCommMessage{
					ID:          getString(m, "id"),
					SenderDID:   getString(m, "sender_did"),
					MessageType: getString(m, "message_type"),
					Payload:     getString(m, "payload"),
					CreatedAt:   getString(m, "created_at"),
				})
			}
		}
	}

	count := getInt(data, "count", len(messages))
	return &DIDCommInbox{
		Messages: messages,
		Count:    count,
	}, nil
}

File: sdks/javascript/src/didcomm.ts
/**
 * DIDComm messaging operations.
 */

import type { ATAPClient } from "./client.js";
import type { DIDCommInbox, DIDCommMessage } from "./models.js";

/** Send and receive DIDComm messages. */
export class DIDCommAPI {
  private readonly _client: ATAPClient;

  constructor(client: ATAPClient) {
    this._client = client;
  }

  /**
   * Send a DIDComm message (JWE envelope). Public endpoint.
   *
   * @param jweBytes - Raw JWE bytes (application/didcomm-encrypted+json).
   * @returns Dict with id and status ("queued").
   */
  async send(jweBytes: Uint8Array): Promise<Record<string, unknown>> {
    return this._client._http.request("POST", "/v1/didcomm", {
      headers: {
        "Content-Type": "application/didcomm-encrypted+json",
      },
      jsonBody: JSON.parse(new TextDecoder().decode(jweBytes)),
    });
  }

  /**
   * Retrieve pending DIDComm messages. Requires atap:inbox scope.
   *
   * @param options.limit - Max messages to return (default 50, max 100).
   */
  async inbox(options: { limit?: number } = {}): Promise<DIDCommInbox> {
    const limit = Math.min(options.limit || 50, 100);
    const data = await this._client._authedRequest(
      "GET",
      "/v1/didcomm/inbox",
      { params: { limit: String(limit) } },
    );
    const messages: DIDCommMessage[] = (
      (data.messages as Record<string, unknown>[]) || []
    ).map((m) => ({
      id: (m.id as string) || "",
      senderDid: (m.sender_did as string) || "",
      messageType: (m.message_type as string) || "",
      payload: (m.payload as string) || "",
      createdAt: (m.created_at as string) || "",
    }));
    return {
      messages,
      count: (data.count as number) || messages.length,
    };
  }
}
